compiler/rustc_data_structures/src/xxhash/mod.rs

```diff
@@ -0,0 +1,1204 @@
+            let mut p = r.start as *const u8;
+            let end = r.end as *const u8;
+
+            write!(f, "{}", tag)?;
+
+            let mut i = 0;
+
+            unsafe {
+                while p != end {
+                    if i % 8 == 7 {
+                        write!(f, "_")?;
+                    }
+
+                    write!(f, "{:02x}", *p)?;
+                    p = p.offset(1);
+                    i += 1;
+                }
+            }
+
+            writeln!(f)
+        }
+
+        writeln!(f, "XXH3State {{")?;
+        write_bytes(f, "  acc          = ", &self.acc.0[..])?;
+        write_bytes(f, "  buffer       = ", &self.buffer.0[..])?;
+
+        writeln!(f, "  num_stripes_so_far = {}", self.num_stripes_so_far)?;
+        writeln!(f, "  total_len = {}", self.total_len)?;
+        writeln!(f, "  num_stripes_per_block = {}", self.num_stripes_per_block)?;
+        writeln!(f, "  secret_limit = {}", self.secret_limit)?;
+        writeln!(f, "  seed         = {}", self.seed)?;
+        writeln!(f, "  buffered_size = {}", self.buffered_size)?;
+        writeln!(f, "}}")?;
+
+        Ok(())
+    }
+}
+
+impl Default for Xxh3Hasher {
+    #[inline]
+    fn default() -> Self {
+        let secret = &XXH3_kSecret.0[..];
+        let secret_limit = secret.len() - XXH_STRIPE_LEN;
+        let num_stripes_per_block = secret_limit / XXH_SECRET_CONSUME_RATE;
+
+        Self {
+            acc: XXH3_INIT_ACC,
+            buffer: /*unsafe {
+                if cfg!(debug_assertions) {
+                    Buffer([0xCC; XXH3_INTERNALBUFFER_SIZE])
+                } else {
+                    std::mem::MaybeUninit::uninit().assume_init()
+                }
+            },*/Buffer([0x00; XXH3_INTERNALBUFFER_SIZE]),
+            buffered_size: 0,
+            num_stripes_so_far: 0,
+            total_len: 0,
+            num_stripes_per_block,
+            secret_limit,
+            seed: 0,
+        }
+    }
+}
+
+impl Hasher for Xxh3Hasher {
+    fn finish(&self) -> u64 {
+        unimplemented!()
+    }
+
+    fn write(&mut self, bytes: &[u8]) {
+        self.update(bytes);
+    }
+}
+
+impl Xxh3Hasher {
+    #[inline]
+    fn secret(&self) -> &[u8] {
+        &XXH3_kSecret.0[..]
+    }
+}
+
+cfg_if! {
+    if #[cfg(target_feature = "sse2")] {
+        const XXH3_accumulate_512: XXH3_f_accumulate_512 = XXH3_accumulate_512_sse2;
+        const XXH3_scrambleAcc: XXH3_f_scrambleAcc = XXH3_scrambleAcc_sse2;
+    } else {
+        const XXH3_accumulate_512: XXH3_f_accumulate_512 = XXH3_accumulate_512_scalar;
+        const XXH3_scrambleAcc: XXH3_f_scrambleAcc = XXH3_scrambleAcc_scalar;
+    }
+}
+
+// Public interface
+impl Xxh3Hasher {
+    // It's important that the fast path, where we just copy things into the state's buffer gets
+    // inlined. It greatly helps performance if the length of the slice is known at compile-time,
+    // so we don't spend a lot of time calling memcpy() for small slices.
```

Comment:
The comment is relevant for both cases I think. It says that it's important to inline this function so that the slice size is known (to the optimizer). It basically explains why there's `#[inline]` below it. `update_fixed_size` then just adds that for this function the "inlining" is guaranteed thanks to const generics.